ui/kotlinx-coroutines-android/test/R8ServiceLoaderOptimizationTest.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.android
+
+import org.jf.baksmali.Adaptors.ClassDefinition
+import org.jf.baksmali.BaksmaliOptions
+import org.jf.dexlib2.DexFileFactory
+import org.jf.dexlib2.iface.ClassDef
+import org.jf.dexlib2.iface.DexFile
+import org.jf.util.IndentingWriter
+import org.junit.Test
+import java.io.File
+import java.io.StringWriter
+import java.util.stream.Collectors
+import kotlin.test.assertEquals
+
+class R8ServiceLoaderOptimizationTest {
+  private val r8Dex = File(System.getProperty("dexPath")!!).asDexFile()
+  private val r8DexNoOptim = File(System.getProperty("noOptimDexPath")!!).asDexFile()
+
+  @Test fun noServiceLoaderCalls() {
+    val serviceLoaderInvocations = r8Dex.classes
+        // Create a map of type names to a list of their ServiceLoader invocations.
+        .associate { clz ->
+          clz.type to clz.toSmali().lines().filter { "Ljava/util/ServiceLoader;->" in it }
```

Comment:
It looks like the only test you are doing here is for textual occurence of the the service loader class. It seems that bringing the whole decompilation library here is a bit of overkill. May I suggest to simply scan the resulting classes for the corresponding string similarly to how we do it in this test: https://github.com/Kotlin/kotlinx.coroutines/blob/master/binary-compatibility-validator/test/PublicApiTest.kt#L78

Reply:
We don't deal with classes in this case, as R8 outputs DEX (and that's what we need to check, as there are differences between optimizations that are applied if I were to switch R8 to output .class files). 
So it looks like being able to parse and read DEX files is essential here. 
I did simplify the test a lot though to not have to generate decompiled source Strings, I simply check if a reference to the type is present in the file.

Reply:
The reason that I chose the smali string representation is that the failure mode is much more useful than when you're querying directly. You're given the actual problems directly rather than having to go investigate separately.